Rework system decorations

ngarside opened the pull request:
## What does the pull request do?
Adds Window.SystemDecorationsProperty allowing three decoration styles:

- Full, system border and titlebar, equivalent to HasSystemDecorations == true
- BorderOnly, system border but no titlebar
- None, no decorations, equivalent to HasSystemDecorations == false

![pr-windows](https://user-images.githubusercontent.com/5486117/74105321-a5d8b400-4b54-11ea-9dcc-455a10bc7d59.png)

![pr-linux](https://user-images.githubusercontent.com/5486117/74105323-a8d3a480-4b54-11ea-8e70-46f9f0254b7b.png)

<img width="512" alt="pr-mac" src="https://user-images.githubusercontent.com/5486117/74105325-aa9d6800-4b54-11ea-9592-16c1dea62882.png">

## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

Commit:
Rework system decorations

kekekeks commented on src/Windows/Avalonia.Win32/WindowImpl.cs:
```diff
@@ -1011,6 +1011,10 @@ private void UpdateWMStyles(Action change)
 
+            MARGINS margins = new MARGINS();
+            margins.cyBottomHeight = _decorated == SystemDecorations.BorderOnly ? 1 : 0;
+            UnmanagedMethods.DwmExtendFrameIntoClientArea(_hwnd, ref margins);
```
What happens on Win7 with a classic theme (aka DWM disabled)?

Symbai replied:
Does Avalonia still (want to) support Win7 now that it has been officially retired?

kekekeks replied:
It still have hundreds of millions of users. By a huge margin more users than Desktop Linux and OSX. Those aren't going anywhere anytime soon. So, no, we aren't planning to drop Win7 support in the immediate future.t

ngarside replied:
I don't have a windows 7 machine to test on, but dwm is only used to add the border for SystemDecorations.BorderOnly, so without dwm it should behave like SystemDecorations.None.

Commit:
Use enum in macOS native

ngarside commented on native/Avalonia.Native/inc/avalonia-native.h:
```diff
@@ -236,7 +236,7 @@ AVNCOM(IAvnWindow, 04) : virtual IAvnWindowBase
     virtual HRESULT ShowDialog (IAvnWindow* parent) = 0;
     virtual HRESULT SetCanResize(bool value) = 0;
-    virtual HRESULT SetHasDecorations(bool value) = 0;
+    virtual HRESULT SetHasDecorations(int value) = 0;
```
Fixed in new commit

ngarside replied:
Updated in new commit

Commit:
Use enum in mac interop

Commit:
Merge branch 'master' into window-decorations

kekekeks commented:
BTW, how would that work with Wayland server decorations once we get the Wayland backend? I only see an on/off toggle in [xdg-decoration](https://github.com/wayland-project/wayland-protocols/blob/master/unstable/xdg-decoration/xdg-decoration-unstable-v1.xml) protocol

ngarside commented:
TBH I don't know Linux all that well so can't say. If I get some time I'll look into how gtk does it when using the header bar.

But even with falling back to None I still think it's a useful feature given how popular the border-without-titlebar style is on Windows and Mac.

MarchingCube commented:
@kekekeks @grokys Can we get a decision on this PR? This is blocking me from resuming my work on owned windows since I will badly conflict with these changes.

Gillibald commented:
Wayland isn't stable yet and I doubt the protocol will be finished any time soon. Also why avoiding a feature just because one technology doesn't support it (yet). You could just ignore unsupported values for now.

kekekeks commented:
@MarchingCube 

I did some investigation regarding xdg_decoration protocol and asked on #wayland IRC channel. It seems that it's a plain ON/OFF switch, so we'll just have to emulate the "BorderOnly" mode by drawing "system" decorations manually (we'll have to do the same for the full decorations on several DEs anyway, so it shouldn't be that much of a problem).
There is also a problem with toggling full decorations ON again after they were disabled, but it should get fixed on the wayland side (I've raised an [issue](https://gitlab.freedesktop.org/wayland/wayland-protocols/issues/9) in wayland-protocols repo).

kekekeks commented:
Regarding the current PR: I'm not sure how future-proof is `SystemDecorations` enum.
```
    SystemDecorationsNone = 0,
    SystemDecorationsBorderOnly = 1,
    SystemDecorationsFull = 2,
```
seems to be very specific. 
Other than that, LGTM.

kekekeks commented on src/Avalonia.Controls/Window.cs:
```diff
@@ -125,7 +153,9 @@ static Window()
             BackgroundProperty.OverrideDefaultValue(typeof(Window), Brushes.White);
             TitleProperty.Changed.AddClassHandler<Window>((s, e) => s.PlatformImpl?.SetTitle((string)e.NewValue));
             HasSystemDecorationsProperty.Changed.AddClassHandler<Window>(
```
It seems that `HasSystemDecorations` and `SystemDecorations` might get out of sync on the `Window`. @grokys any idea how to properly synchronize those?

grokys replied:
Do you mean `HasSystemDecorations` and the value on the platform impl? The pattern we're using elsewhere is exposing a property and a changed `Action` on the impl.

(I think we should use another mechanism but that's a different subject)

kekekeks replied:
There are now two avalonia properties that set the same platform property. Those will be out of sync.

grokys replied:
Ah, got you. Yeah, that's a bit of a difficult situation because both are styled properties. Perhaps we should change `HasSystemDecorations` to a direct property that sets `SystemDecorations`? It'd be a breaking change in the definition of `HasSystemDecorationsProperty` but not one that should cause a huge problem I hope?

We'd then define `HasSystemDecorations` as:

```
public bool HasSystemDecorations
{
    get => SystemDecorations != SystemDecorations.None;
    set => SystemDecorations = value ? SystemDecorations.Full : SystemDecorations.None;
}
```

grokys replied:
BTW, we talked a while ago about emitting warnings for obsolete properties from the XAML compiler. Would that be that difficult to do? Would help warn people about situations like this.

kekekeks replied:
There could be some existing code that uses data binding to `HasSystemDecorations` property.

grokys replied:
That should continue working. We'd have to raise `RaisePropertyChanged` on `HasSystemDecorations` when `SystemDecorations` changes but bindings should continue working as before.

ngarside replied:
I've added a commit which fixes this but not sure if it's the preferred way?

@grokys suggestion is cleaner but I'm not sure how to implement it with avalonia's properties i.e. how to bind it so that setting SystemDecorations causes a property changed event for HasSystemDecorations and vice versa?

ngarside commented:
Sorry for they delay, I haven't had much time to work on this lately. Thanks for checking wayland @kekekeks. I'll look into fixing the HasSystemDecorations property.

> Regarding the current PR: I'm not sure how future-proof is `SystemDecorations` enum.
> 
> ```
>     SystemDecorationsNone = 0,
>     SystemDecorationsBorderOnly = 1,
>     SystemDecorationsFull = 2,
> ```
> 
> seems to be very specific.
> Other than that, LGTM.

I did consider using flags for the enum, that way more options could be added in the future (e.g. border without shadow, or windowbuttons without titlebar). But then you run into problems of different combinations not working on different platforms and some not working at all and it all gets a bit messy.

None/Border/Full seemed like a good balance of usefulness and cross platform support. Open to changing it though if you'd prefer a different approach.

```csharp
[Flags]
public enum SystemDecorations
{
    None = 0
    Border = 1
    Titlebar = 2
    Full = Border + Titlebar
}
```

Commit:
Fix event binding for HasSystemDecorations property

kekekeks commented:
>    Full = Border + Titlebar

Titlebar without border doesn't make much sense though

BTW, It seems that on OSX the border flag controls the shadow rather than border.

ngarside commented:
It was an idea to make it more future proof, but yes I agree it's not a great solution.

Testing on macos 10.15 it adds the rounded corners and border for me. Is that not what you're getting?

grokys committed:
Merge branch 'master' into window-decorations

grokys committed:
Rework HasSystemDecorations.

Make it a direct property that is backed by `SystemDecorations`. Moved the changed handlers to `OnPropertyChanged`.

grokys commented:
Apologies for the delay @ngarside - I've just pushed a commit to your branch which implements `HasSystemDecorations` in terms of `SystemDecorations`, hope you don't mind me pushing directly to your branch!

@kekekeks how's this looking now? Ready for merge?

kekekeks commented:
Still not really convinced about the API design, other than that LGTM.

We might want to change that to use "hint" flags before 0.10 release

grokys commented:
Any suggestions for a better API design?

kekekeks commented:
>If this is an relatively straightforward change

It would only affect the `Window.cs`, the rest would be the same.

I was thinking about something like `WindowHints.KeepBorderWhenDecorationsAreDisabled`, but it looks kinda ugly too.

ngarside commented:
Thanks @grokys.

@kekekeks window hints would just be flags on the window right? I can update this pr if you're decided on that API design?

kekekeks commented:
I'm not quite sure about using flags since those aren't really XAML and data binding friendly. At the same time I don't want to pollute the list of Window properties. May be we should use something like
`WindowDecorationHints.KeepBorderWhenDecorationsAreDisabled` attached property?
@grokys

grokys commented:
Sorry @kekekeks without a more detailed explanation of what you have in mind I can't really comment. Lets get this merged anyway and work from there.

grokys committed:
Merge branch 'master' into window-decorations

Brouilles commented:
SystemDecorations is in 0.9.6? 
I can't use it in my project.

Thanks.

ngarside commented:
It's not in 0.9.x, should be in 0.10.

Gillibald commented:
New features are only available threw https://github.com/AvaloniaUI/Avalonia/wiki/Using-nightly-build-feed

Brouilles commented:
Thanks. Sorry for the convenient.

Enoq00 commented:
Is it in 0.10.
I update to 0.10 but still can’t use it.
Thank you.

ngarside commented:
Are you sure you're on 0.10, I can't find a reference to its release anywhere

Enoq00 commented:
I got it from nuget

ngarside commented:
Link? The latest release I can see on nuget is 0.9.10, which this won't be in as 0.9.x point releases are only for bug fixes.

If you need this feature now you'd have to use Gillibald's link above.

Enoq00 commented:
Sorry for careless. Yes im on 0.9.10
You know when 0.10 will release?

ngarside commented:
I don't know sorry, you'd have to ask @kekekeks or one of the other maintainers.

Gillibald commented:
Feel free to use the nightly build feed http://avaloniaui.net/docs/quickstart/packages
That way you get most recent changes
